### PR TITLE
Fixing Database tests and Snowflake Dialect - part 2 out of ...

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Part of #9486 
- Fixing our tests to not rely on deterministic ordering of created Tables in Database backends
	- Before, SQLite and Postgres used to mostly return rows in the order they were inserted in, but Snowflake does not.
- Fixing various parts of Snowflake dialect.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: test/Table_Tests/src/Common_Table_Operations/Add_Row_Number_Spec.enso [hunk 1]
```diff
@@ -84,7 +78,7 @@ add_specs suite_builder setup =
 
             t3 = t.add_row_number order_by=["X"] |> materialize |> _.order_by "row_id"
             Problems.assume_no_problems t3
-            t3.at "Row" . to_vector . should_equal [1, 3, 2, 4, 5]
+            t3.at "Row" . to_vector . should_have_relative_ordering [[1, 3], [2], [4, 5]]
```
Comment (radeusgd) [inline test/Table_Tests/src/Common_Table_Operations/Add_Row_Number_Spec.enso:81]: This relative ordering is according to spec - the elements that are equal on `X` can be in arbitrary order, only elements with differing `X` values need to be monotonic.

We could technically make this more detailed test in in-memory - requiring that the sort is 'stable' - i.e. the order of elements of the same equivalence class is the same as in the input. What do you think @AdRiley ? If you think it's useful I can add it in later PR.

File: distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso [hunk 2]
```diff
@@ -364,7 +364,9 @@ agg_count_distinct args = if args.is_empty then (Error.throw (Illegal_Argument.E
 ## PRIVATE
 agg_count_distinct_include_null args =
     ## If we always count as tuples, then even null fields are counted.
-    SQL_Builder.code "COUNT(DISTINCT (" ++ SQL_Builder.join ", " args ++ ", 0))"
+    ## But Snowflake seems to not like tuples?
+    #SQL_Builder.code "COUNT(DISTINCT (" ++ SQL_Builder.join ", " args ++ ", 0))"
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso:367]: Is there a TODO to restore this previous verison?

File: test/Snowflake_Tests/src/Snowflake_Spec.enso [hunk 3]
```diff
@@ -482,76 +506,78 @@ snowflake_specific_spec suite_builder create_connection_fn db_name setup =
...
-
+        in_mem_table.select_into_database_table (connection.if_nothing default_connection.get) name primary_key=Nothing temporary=True
+    light_table_builder columns =
+        default_connection.get.base_connection.create_literal_table (Table.new columns) "literal_table"
     materialize = .read
 
     Common_Spec.add_specs suite_builder prefix create_connection_fn
 
-    common_selection = Common_Table_Operations.Main.Test_Selection.Config supports_case_sensitive_columns=True order_by_unicode_normalization_by_default=True allows_mixed_type_comparisons=False fixed_length_text_columns=True removes_trailing_whitespace_casting_from_char_to_varchar=True supports_decimal_type=True supported_replace_params=supported_replace_params
+    common_selection = Common_Table_Operations.Main.Test_Selection.Config supports_case_sensitive_columns=True order_by_unicode_normalization_by_default=True allows_mixed_type_comparisons=False fixed_length_text_columns=True removes_trailing_whitespace_casting_from_char_to_varchar=True supports_decimal_type=True supported_replace_params=supported_replace_params run_advanced_edge_case_tests_by_default=False
```
Comment (GregoryTravis) [inline test/Snowflake_Tests/src/Snowflake_Spec.enso:570]: Is there a particular reason not to run the advanced edge cases? Are they slow?

**Review by GregoryTravis — APPROVED** [review]

**Review by hubertp — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso [hunk 4]
```diff
@@ -0,0 +1,9 @@
+## PRIVATE
+   A simple type encapsulating the pattern of delaying initialization of some
+   value.
+type Lazy_Ref
+    ## PRIVATE
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso]: Can we use `private`?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso [hunk 5]
```diff
@@ -0,0 +1,9 @@
+## PRIVATE
+   A simple type encapsulating the pattern of delaying initialization of some
+   value.
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso]: Why do we need `Lazy_Ref` when the same functionality is part of the language?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso [hunk 6]
```diff
@@ -0,0 +1,9 @@
+## PRIVATE
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso]: Make the module `private`.

File: distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso [hunk 7]
```diff
@@ -12,8 +12,10 @@ type Group_Builder
     ## PRIVATE
 
        Arguments:
+       - name: The name of the group.
+       - builder: Vector builder used for storing specs.
        - teardown_ref: A reference to a teardown method.
-    Impl (builder = Builder.new) (teardown_ref = Ref.new (_ -> Nothing))
+    Impl (name : Text) (builder = Builder.new) (teardown_ref = Ref.new (_ -> Nothing))
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso:18]: Can the constructor be made `private`?

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
Please remove `Lazy_Ref`.

What's the point of polluting `Standard.Base` API with `## PRIVATE` utility that is so simple, it can be realized with a trivial language construct in any module that needs it? DRY? I don't believe in DRY when designing APIs. I value encapsulation way more.

I am requesting changes, but should there be a consensus that `##PRIVATE` `Lazy_Ref` belongs to `Standard.Base`, I am ready to change my review to comment to unblock the PR.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso (hunk 6 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso]: Module can't be private as used in Table_Tests.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+   A simple type encapsulating the pattern of delaying initialization of some
+   value.
+type Lazy_Ref
+    ## PRIVATE
+    Value ~get
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso]:
```suggestion
    ## PRIVATE
    private Value ~internal_get
    
    ## Materialises the value and returns the result
    get self = self.internal_get
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso (hunk 6 above)
Comment (JaroslavTulach) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso]: Aren't tests supposed to run with `--disable-private-check`?

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso]: No, at least not right now. We need to design a sensible approach to that.

If we run all tests with `--disable-private-check` we will not catch regressions where one library uses another's `private` entity. In tests with `--disable-private-check` such usage will succeed, but then in production it would fail.

It's not an imagined scenario - while refactoring I already encountered it numerous times where I made "too many things private" and things stopped working. If tests are ran with this flag without some further design, we will risk breaking too easily.

IMO we could add a separate project like `Internals_Tests`, or for each project have e.g. `Table_Internals_Tests`. Such test project could run with `--disable-private-check`. But the main suite cannot, at least not in the current setup.

File: distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso (hunk 7 above)
Comment (radeusgd) [reply distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso:18]: Probably.

File: test/Snowflake_Tests/src/Snowflake_Spec.enso (hunk 3 above)
Comment (radeusgd) [reply test/Snowflake_Tests/src/Snowflake_Spec.enso:570]: Yes, exactly. We are running **thousands** of queries in the tests. In Postgres each query runs ~1ms so that's maybe acceptable but for Snowflake where queries are closer to 100ms mark - this makes the whole suite run for about an hour. They can be ran on a special nightly run that is OK if running for a longer time.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso (hunk 5 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso]: Currently in Enso, I cannot create a lazy _value_. 

If you give me a way to initialize an inner value lazily, e.g.:
```
my_function group_builder x =
    ~lazy_val = expensive_computation x
    group_builder.specify "some test" <|
        lazy_val . should_equal 42
    group_builder.specify "some other test" <|
        lazy_val . should_be_a Integer
```
than I will not need `Lazy_Ref`.

Until then, every time I need a pattern like the above, I need to re-implement `Lazy_Ref`. I wanted to have this common functionality so that I can just use it, instead of having to copy-paste it every time I need it.

But if you are really strongly opposed to it I will just keep wasting time copy-pasting.

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy_Ref.enso]: I have removed the helper as per your request and I'm just adding the `type Lazy_Ref` wherever it is used.
Please approve the PR, so that I can proceed.

File: distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso:367]: I'm pretty sure we will need to fix handling it properly to make our tests pass. So I think no TODO shall be needed.

File: distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso (hunk 7 above)
Comment (radeusgd) [reply distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso:18]: I have tried this, but it broke the benchmarks.

image


Modifying privacy of the test constructors is completely unrelated to these changes, so I reverted it.

**Review by JaroslavTulach — APPROVED** [review]
